tooling/greybox_fuzzer/src/mutation/mod.rs

```diff
@@ -0,0 +1,769 @@
+use acvm::{AcirField, FieldElement};
+
+use array::{mutate_vector_structure, splice_array_structure};
+use configurations::{
+    TestCaseSpliceType, BASIC_TESTCASE_SPLICE_CONFIGURATION,
+    BASIC_TOP_LEVEL_MUTATION_CONFIGURATION, BASIC_UNBALANCED_ARRAY_SPLICE_MUTATION_CONFIGURATION,
+};
+use dictionary::FullDictionary;
+use field::mutate_field_input_value;
+use int::mutate_int_input_value;
+use noirc_abi::{input_parser::InputValue, Abi, AbiType, InputMap};
+use rand::Rng;
+use rand_xorshift::XorShiftRng;
+use std::{
+    collections::{BTreeMap, HashSet},
+    iter::zip,
+};
+use string::{mutate_string_input_value, splice_string_input_value};
+
+mod array;
+mod configurations;
+mod dictionary;
+mod field;
+mod int;
+mod string;
+use num_traits::Zero;
+pub struct InputMutator {
+    abi: Abi,
+    weight_tree: NodeWeight,
+    full_dictionary: FullDictionary,
+}
+
+const MUTATION_LOG_MIN: u32 = 0;
+const MUTATION_LOG_MAX: u32 = 5;
+/// NodeWeight determines the probability of mutating a particular object
+/// It represent the relative weight of this InputValue and its children
+#[derive(Clone, Debug)]
+pub struct NodeWeight {
+    start: u32,
+    end: u32,
+    subnodes: Option<Vec<NodeWeight>>,
+}
+impl NodeWeight {
+    pub fn get_weight(&self) -> u32 {
+        self.end - self.start
+    }
+    pub fn calculate_offsets(&mut self, start_offset: u32) {
+        assert!(self.start == 0, "Only calculate offsets after initial computation");
+        self.start += start_offset;
+        self.end += start_offset;
+        let mut current_update = start_offset;
+        if self.subnodes.is_some() {
+            let subnode_weights = self.subnodes.as_mut().unwrap();
+            for subnode_weight in subnode_weights.iter_mut() {
+                let weight_update = subnode_weight.get_weight();
+                subnode_weight.calculate_offsets(current_update);
+                current_update += weight_update;
+            }
+        }
+    }
+}
+
+enum UnbalancedSplice {
+    FirstTestCase,
+    SecondTestCase,
+}
+
+struct UnbalancedSpliceConfiguration {
+    first_testcase_weight: usize,
+    #[allow(unused)]
+    second_testcase_weight: usize,
+    total_weight: usize,
+}
+
+impl UnbalancedSpliceConfiguration {
+    #[allow(unused)]
+    pub fn new(first_testcase_weight: usize, second_testcase_weight: usize) -> Self {
+        let total_weight = first_testcase_weight + second_testcase_weight;
+        Self { first_testcase_weight, second_testcase_weight, total_weight }
+    }
```

Comment:
Can this be removed if it's unused? Same goes for `second_testcase_weight`.

Reply:
Moved it into configurations